src/journal/compress.c

```diff
@@ -669,20 +914,118 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
                   total_in, total_out,
                   total_in > 0 ? (double) total_out / total_in * 100 : 0.0);
  cleanup:
-        munmap(src, st.st_size);
+        (void) munmap(src, st.st_size);
         return r;
 #else
         log_debug("Cannot decompress file. Compiled without LZ4 support.");
         return -EPROTONOSUPPORT;
 #endif
 }
 
+int decompress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeDCtxp) ZSTD_DCtx *s = NULL;
+        size_t ret;
+        uint64_t left = max_bytes, in_bytes = 0;
+        int r;
+        ssize_t ss;
+        size_t out_size, in_size;
+        _cleanup_free_ char *out;
+        char *in;
+        ZSTD_outBuffer ZSTD_out;
+        ZSTD_inBuffer ZSTD_in;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        s = ZSTD_createDStream();
+        if (!s)
+                return -ENOMEM;
+        out_size = ZSTD_DStreamOutSize();
+        in_size = ZSTD_DStreamInSize();
+        out = malloc(out_size + in_size);
+        if (!out)
+                return -ENOMEM;
+        in = out + out_size;
+        ZSTD_in.src = in;
+        ZSTD_in.size = in_size;
+        ZSTD_in.pos = 0;
+        ZSTD_out.dst = out;
+        ZSTD_out.size = out_size;
+        ZSTD_out.pos = 0;
+
+        ret = ZSTD_initDStream(s);
+        if (ZSTD_isError(ret) > 0) {
+                log_debug("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errno(ret);
+        }
+
+        for (;;) {
+                bool first = true;
+
+                ss = read(fdf, in, in_size);
+                if (ss < 0)
+                        return -errno;
+                if (ss == 0)
+                        break;
+                ZSTD_in.size = ss;
+                in_bytes += ss;
+
+                if (first) {
+                        unsigned long long q;
+
+                        q = ZSTD_getFrameContentSize(ZSTD_in.src, ZSTD_in.size);
+                        if (q == ZSTD_CONTENTSIZE_UNKNOWN)
+                                ;
+                        else if (q == ZSTD_CONTENTSIZE_ERROR)
+                                return -EBADMSG;
+                        else if (q > max_bytes)
+                                return -EFBIG;
+                        first = false;
+                }
+
+                while (ZSTD_in.pos < ZSTD_in.size) {
```

Comment:
I somehow managed to trigger an infinity loop by running the following commands:
```sh
sh -c 'kill -ABRT $$'
./build/coredumpctl dump -1 | zstd -o CORE.zst
printf "MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1\nMESSAGE=CORE\nCOREDUMP_FILENAME=$(pwd)/CORE.zst\n" | logger --journald
./build/coredumpctl dump -1 -o CORE
```
I used `libzstd-1.3.3-1.fc27.x86_64`.

Reply:
I haven't figured out what was going on yet, but I've just realized that I had to send the message containing `COREDUMP_FILE` to the journal manually because `systemd-coredump` currently can't do it. I think the phrase "Closes: #8097" should be removed for now so that #8097 won't be closed automatically when the PR is merged. Ideally, `systemd-coredump` should be able to compress coredumps via `libzstd`, but my understanding of this PR (which may be totally incorrect) is that it isn't supposed to implement everything at once.